src/Glimpse.Server.Web/Storage/InMemoryStorage.cs

```diff
@@ -8,65 +7,200 @@ namespace Glimpse.Server.Web
 {
     public class InMemoryStorage : IStorage, IQueryRequests
     {
+        /// <summary>
+        /// Internal class to contain all data associated with a single request. 
+        /// </summary>
+        private class RequestInfo
+        {
+            public RequestInfo(LinkedListNode<Guid> requestIdNode)
+            {
+                this.RequestIdNode = requestIdNode;
+                this.Messages = new LinkedList<IMessage>();
+                this.Indices = null;
+            }
+
+            /// <summary>
+            /// List of individual messages associated with this request
+            /// </summary>
+            public LinkedList<IMessage> Messages { get; private set; }
+            
+            /// <summary>
+            /// The node in the _activeRequests list.  Storing this here allows us to determine to track 
+            /// requests by age in constant time.
+            /// </summary>
+            public LinkedListNode<Guid> RequestIdNode { get; private set; }
+
+            /// <summary>
+            /// Indices associated with this Request
+            /// </summary>
+            public RequestIndices Indices { get; set; }
+        }
+
         public const int RequestsPerPage = 50;
-        public const int MaxRequests = 500; // TODO: Make this configurable
+        public const int DefaultMaxRequests = 500;
 
-        private readonly List<IMessage> _messages;
-        private readonly ConcurrentDictionary<Guid, RequestIndices> _indices;
+        /// <summary>
+        /// Primary storage for Messages. 
+        /// </summary>
+        private readonly Dictionary<Guid, RequestInfo> _requestTable;
 
-        private static readonly object _locker = new object(); // HACK :(
+        /// <summary>
+        /// Currently active requests, ordered by incoming messages for the request.  The first entry in the list will be
+        /// the most recent request to receive a message, the last entry the oldest request to receive a message.
+        /// </summary>
+        private readonly LinkedList<Guid> _activeRequests;
```

Comment:
I can add context around the `TriggerCleanup()`:

For *this* implementation of `IStorage`, I decided not to let users decide how many messages were stored, but rather how many requests. I did this because:

1. Will users know how many messages a request generates? That number will vary based on the specific request and the number of extensions the user has installed.
2. If they did specify a number of message to be stored, we'd most likely end up with many "half requests" in storage, and the UI would seem broken for those requests.

I figured the user would be able to reason about requests a little bit more. It's true that gives them less granularity in controlling Glimpse - but since they don't control message sizes, that control is a bit of an illusion anyways.